File: Ahorn/entities/ShieldedRefill.jl
module CommunalHelperShieldedRefill
using ..Ahorn, Maple

@mapdef Entity "CommunalHelper/ShieldedRefill" ShieldedRefill(x::Integer, y::Integer, 
	twoDashes::Bool = false, oneUse::Bool = false, bubbleRepel::Bool = true)    

const placements = Ahorn.PlacementDict(
    "Shielded Refill (Two Dashes) (Communal Helper)" => Ahorn.EntityPlacement(
        ShieldedRefill,
        "point",
        Dict{String, Any}(
            "twoDashes" => true
        )
    ),
    "Shielded Refill (Communal Helper)" => Ahorn.EntityPlacement(
        ShieldedRefill,
        "point",
        Dict{String, Any}(
            "twoDashes" => false
        )
    )    
)

const spriteOneDash = "objects/refill/idle00"
const spriteTwoDash = "objects/refillTwo/idle00"

getSprite(entity::ShieldedRefill) = get(entity.data, "twoDashes", false) ? spriteTwoDash : spriteOneDash

function Ahorn.selection(entity::ShieldedRefill)
    x, y = Ahorn.position(entity)
    sprite = getSprite(entity)

    return Ahorn.getSpriteRectangle(sprite, x, y)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::ShieldedRefill)
    Ahorn.Cairo.save(ctx)
    Ahorn.set_antialias(ctx, 1)
    Ahorn.set_line_width(ctx, 1);

    Ahorn.drawCircle(ctx, 0, 0, 8, (1.0, 1.0, 1.0, 1.0))
    Ahorn.Cairo.restore(ctx)

    sprite = getSprite(entity)
    Ahorn.drawSprite(ctx, sprite, 0, 0)
end

end

File: Loenn/entities/chains/shielded_refill.lua
local drawableSprite = require("structs.drawable_sprite")

local shieldedRefill = {}

shieldedRefill.name = "CommunalHelper/ShieldedRefill"
shieldedRefill.depth = -100
shieldedRefill.placements = {
    {
        name = "one_dash",
        data = {
            twoDashes = false,
            oneUse = false,
            bubbleRepel = true
        }
    },
    {
        name = "two_dashes",
        data = {
            twoDashes = true,
            oneUse = false,
            bubbleRepel = true
        }
    }
}

local function getSprite(entity)
    return drawableSprite.fromTexture(
        entity.twoDashes and "objects/refillTwo/idle00" or "objects/refill/idle00",
        entity
    )
end

function shieldedRefill.selection(room, entity)
    local sprite = getSprite(entity)

    return sprite:getRectangle()
end

function shieldedRefill.draw(room, entity, viewport)
    local x, y = entity.x or 0, entity.y or 0

    love.graphics.circle("line", x, y, 8)
    getSprite(entity):draw()
end

return shieldedRefill
